Improve and harden router deletion on agent

sebageek opened the pull request:
This is not yet a fix for the agent unscheduling problem, but this is at least half the way there.

sebageek committed:
Improve and harden router deletion on agent

When delete_router() in neutron is called it gets deleted from the
database. Afterwards all agents are called with a router_deleted()
notification. This puts a router in the processing queue, which then
tries to get router data from neutron - which is no longer there.

To get the router data and do a proper delete (i.e. not let the
scavanger do the work) a copy of the router config is saved in the
neutron cache (which times out after 10 minutes by default). When an
agent is then asking for router config of a deleted router via
get_deleted_router() the result is taken from said cache (keyed by
router id and host).

joker-at-work commented on asr1k_neutron_l3/common/cache_utils.py:
```diff
@@ -0,0 +1,57 @@
+    if not cache:
+        return
+
+    key = _gen_cache_key("deleted-router", host, router_id)
```
Would probably make sense to move this prefix into a variable as it's used in two places.

joker-at-work commented on asr1k_neutron_l3/plugins/l3/rpc/rpc_api.py:
```diff
@@ -55,11 +56,12 @@ def update_router_status(self, context, **kwargs):
+        router = cache_utils.get_deleted_router(host, router_id)
+        if not router:
+            LOG.debug("No entry for router %s on host %s in deleted router cache", router_id, host)
+            return None
```
This `None` would not be necessary.

joker-at-work reviewed (commented):
In general, do we expect any problems with serializing the data of the router into the cache? Are there any objects, that aren't used too often, that might fail the serialization?

sebageek commented:
From what I've seen the output of `get_sync_data()` only contains dicts, lists, strings and numbers, so serializing should not be a problem. I didn't follow all code paths though. On the other hand this datastructure at least gets serialized for transport through the rpc framework from neutron to the asr1k agent.

joker-at-work reviewed (approved):
If it's going over the wire, then neutron probably removed all the problematic objects or has compatibility shims for serialization. :+1:

sebageek committed:
Don't update already deleted routers

The router processing queue might still have a router in it scheduled
for update (with the router data already present), which might sync us
an already deleted router back on the device. To prevent this we keep a
cache of deleted router ids for about an hour. If a router is
unscheduled and then rescheduled or updated in any other way after the
delete then we remove it from this cache.
